cmd/ci-secret-bootstrap/main.go

```diff
@@ -994,11 +1231,292 @@ func reconcileSecrets(o options, client secrets.ReadOnlyClient, prowDisabledClus
 			errs = append(errs, fmt.Errorf("failed to write secrets on dry run: %w", err))
 		}
 	} else {
-		if err := updateSecrets(o.secretsGetters, secretsMap, o.force, o.confirm, sets.New[string](o.config.OSDGlobalPullSecretGroup()...), prowDisabledClusters); err != nil {
+		if err := updateSecrets(o.secretsGetters, secretsMap, o.force, o.confirm, sets.New[string](o.vaultConfig.OSDGlobalPullSecretGroup()...), prowDisabledClusters); err != nil {
 			errs = append(errs, fmt.Errorf("failed to update secrets: %w", err))
 		}
 		logrus.Info("Updated secrets.")
 	}
 
 	return errs
 }
+
+// mergeSecretMaps combines Vault and GSM secret maps, with Vault taking precedence on conflicts.
+// Returns the merged map and any conflict errors encountered.
+func mergeSecretMaps(vaultSecrets, gsmSecrets map[string][]*coreapi.Secret) (map[string][]*coreapi.Secret, error) {
+	if len(gsmSecrets) == 0 {
+		return vaultSecrets, nil
+	}
+	if len(vaultSecrets) == 0 {
+		return gsmSecrets, nil
+	}
+
+	vaultIndex := make(map[string]map[types.NamespacedName]bool)
+	for cluster, secretList := range vaultSecrets {
+		if vaultIndex[cluster] == nil {
+			vaultIndex[cluster] = make(map[types.NamespacedName]bool)
+		}
+		for _, secret := range secretList {
+			nsName := types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}
+			vaultIndex[cluster][nsName] = true
+		}
+	}
+
+	var errs []error
+	merged := make(map[string][]*coreapi.Secret)
+
+	for cluster, secretList := range vaultSecrets {
+		merged[cluster] = make([]*coreapi.Secret, len(secretList))
+		copy(merged[cluster], secretList)
+	}
+
+	for cluster, secretList := range gsmSecrets {
+		for _, secret := range secretList {
+			nsName := types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}
+
+			if vaultIndex[cluster] != nil && vaultIndex[cluster][nsName] {
+				errs = append(errs, fmt.Errorf(
+					"conflict: GSM secret %s/%s on cluster %s conflicts with Vault (Vault takes precedence)",
+					secret.Namespace, secret.Name, cluster,
+				))
+				continue
+			}
+
+			merged[cluster] = append(merged[cluster], secret)
+		}
+	}
+
+	return merged, utilerrors.NewAggregate(errs)
+}
+
+// collectionGroupKey is used to track auto-discovered fields for a collection+group pair
+type collectionGroupKey struct {
+	collection string
+	group      string
+}
+
+// constructSecretsFromGSM fetches secrets from GSM and builds Kubernetes Secret objects.
+// For bundles without explicit field lists, it discovers fields by querying GSM.
+// Returns a map of cluster name to list of Kubernetes Secret objects, and any fetch/build errors.
+func constructSecretsFromGSM(
+	ctx context.Context,
+	gsmConfig secretbootstrap.GSMConfig,
+	gsmClient gsm.SecretManagerClient,
+	gsmProjectConfig gsm.Config,
+	prowDisabledClusters sets.Set[string]) (map[string][]*coreapi.Secret, error) {
+	var errs []error
+	uniqueSecretNames := make(map[gsmSecretRef]bool)
```

Comment:
I guess that a `make(map[gsmSecretRef]struct{})` is more convenient here. The `struct{}` construct allocates no memory in Golang.

Update: or maybe use `sets.Set`.